apiplugin/file/file_test.go

```diff
@@ -1,26 +1,193 @@
+	"github.com/iansmith/parigot/g/file/v1"
+)
 
 // if you look at the tests in queue_test you'll see that
 // the tests ignore the wrapper functions and use the
 // real impleentations directly.
 
+const fileContent = "Hello! Parigot!"
+const filePath = "/parigot/app/file.go"
+
 func TestOpenClose(t *testing.T) {
+	svc := newFileSvc((context.Background()))
+	creatAGoodFile(t, svc)
 
-	// try open a file with a badly formed path name, make sure it fails
-	// see filepath.Clean()
+	// try opening a file does not exist
+	badFid := testFileOpen(t, svc, "/parigot/app/badfile.txt", "path doesn't exist",
+		true, int32(file.FileErr_NotExistError))
+	if !badFid.IsEmptyValue() {
+		t.Errorf("trying to open a file which does not exist")
+	}
+	// try open a file with a badly formed path name
+	badFid = testFileOpen(t, svc, "badfile.txt", "bad path name",
+		true, int32(file.FileErr_InvalidPathError))
+	if !badFid.IsEmptyValue() {
+		t.Errorf("accidentally opened a file with the a bad path name")
+	}
 
 	// try opening and closing a file
+	fid := testFileOpen(t, svc, filePath, "open a file", false, int32(file.FileErr_NoError))
+	testFileClose(t, svc, fid, "close a file", false, int32(file.FileErr_NoError))
 
 	// also try opening a file twice, for now should be an error
-	// also try closing a file twice
-
-	// when expecting an error, for now you'll have to use
-	// the raw numbers from the list of errors
-	// apiwasm/file/fileerr.go
+	creatAGoodFile(t, svc)
+	fid = testFileOpen(t, svc, filePath, "open a file", false, int32(file.FileErr_NoError))
+	testFileOpen(t, svc, filePath, "open a open file", true, int32(file.FileErr_PermissionError))
+	// also try closing a file twice, there should be an error in the second time
+	testFileClose(t, svc, fid, "close a file", false, int32(file.FileErr_NoError))
+	testFileClose(t, svc, fid, "close a file that does not exist", true, int32(file.FileErr_NotExistError))
 }
 
 func TestCreateClose(t *testing.T) {
+	svc := newFileSvc((context.Background()))
+
+	// create a file with duplicate "/" in the path name
+	badFid := testFileCreate(t, svc, "/parigot/app///file.go", fileContent,
+		"bad path name", true, int32(file.FileErr_InvalidPathError))
+	if !badFid.IsZeroValue() {
+		t.Errorf("accidentally created a file with the a bad path name")
+	}
+	// create a file with . in the path name
+	badFid = testFileCreate(t, svc, "/parigot/app/./file.go", fileContent,
+		"bad path name", true, int32(file.FileErr_InvalidPathError))
+	if !badFid.IsZeroValue() {
+		t.Errorf("accidentally created a file with the a bad path name")
+	}
+	// create a file with .. in the path name
+	badFid = testFileCreate(t, svc, "/parigot/app/../file.go", fileContent,
+		"bad path name", true, int32(file.FileErr_InvalidPathError))
+	if !badFid.IsZeroValue() {
+		t.Errorf("accidentally created a file with the a bad path name")
+	}
+	// create a file without prefix '/parigot/app/'
+	badFid = testFileCreate(t, svc, "dir/file.go", fileContent, "bad path name",
```

Comment:
Might want to test this with a filename like `/parigot/application/` or something else that is "close" to being right.